PagingWithNetworkSample/app/src/main/java/com/android/example/paging/pagingwithnetwork/reddit/repository/inDb/PageKeyedRemoteMediator.kt

```diff
@@ -0,0 +1,90 @@
+import com.android.example.paging.pagingwithnetwork.reddit.db.RedditDb
+import com.android.example.paging.pagingwithnetwork.reddit.db.RedditPostDao
+import com.android.example.paging.pagingwithnetwork.reddit.db.RedditPostRemoteKeyDao
+import com.android.example.paging.pagingwithnetwork.reddit.vo.RedditPost
+import com.android.example.paging.pagingwithnetwork.reddit.vo.RedditPostRemoteKey
+import retrofit2.HttpException
+import java.io.IOException
+
+@OptIn(ExperimentalPagingApi::class)
+class PageKeyedRemoteMediator(
+    private val db: RedditDb,
+    private val redditApi: RedditApi,
+    private val subredditName: String
+) : RemoteMediator<Int, RedditPost>() {
+    private val postDao: RedditPostDao = db.posts()
+    private val remoteKeyDao: RedditPostRemoteKeyDao = db.remoteKeys()
+
+    override suspend fun load(
+        loadType: LoadType,
+        state: PagingState<Int, RedditPost>
+    ): MediatorResult {
+        try {
+            // Get the closest item from PagingState that we want to load data around.
+            val loadKey = when (loadType) {
+                REFRESH -> null
+                PREPEND -> return MediatorResult.Success(endOfPaginationReached = true)
+                APPEND -> {
+                    val redditPost = state.pages
+                        .lastOrNull { it.data.isNotEmpty() }
+                        ?.data?.lastOrNull()
+
+                    // Query DB for RedditPostPageKey corresponding to the item above.
+                    // RedditPostPageKey is a wrapper object we use to keep track of page keys we
+                    // receive from the Reddit API to fetch the next or previous page.
+                    val postKey = db.withTransaction {
+                        redditPost?.let { remoteKeyDao.remoteKeyByPost(it.name, it.subreddit) }
+                    }
+
+                    // We must explicitly check if the page key is null when appending, since the
+                    // Reddit API informs the end of the list by returning null for page key, but
+                    // passing a null key to Reddit API will fetch the initial page.
+                    if (postKey == null) {
+                        return MediatorResult.Success(endOfPaginationReached = true)
+                    }
+
+                    postKey.nextPageKey
+                }
+            }
+
+            val data = redditApi.getTop(
+                subreddit = subredditName,
+                after = loadKey,
+                before = null,
+                limit = if (loadType == REFRESH) state.config.initialLoadSize else state.config.pageSize
+            ).data
+
+            val items = data.children.map { it.data }
+
+            db.withTransaction {
+                if (loadType == REFRESH) {
+                    postDao.deleteBySubreddit(subredditName)
+                    remoteKeyDao.deleteBySubreddit(subredditName)
+                }
+
+                remoteKeyDao.insertAll(
```

Comment:
this seems unnecessary.

we don't have local modifications here.

All we need to do is:

ignore the paging state, we know it only calls us after we are done w/ local data.

Keep 1 item in db per subreddit

```
SubredditPageKey(
   name : String,
   prevKey : String,
   nextKey : String
)
```

when request comes, load the key item and make the request based on that. Also just keep a global before & after keys (we actually probably only need after key).

Since we also only load in 1 direction i don't think we need to sync on that.

Reply:
so here is the pseudo code:

```
@Entity
SubredditPageInfo(
   name : String,
   nextKey : String
)

override suspend fun load(
        loadType: LoadType,
        state: PagingState<Int, RedditPost>
    ): MediatorResult {
    val pageInfo : SubredditPageInfo? = when (loadType) {
                REFRESH -> null
                PREPEND -> return MediatorResult.Success(endOfPaginationReached = true)
                APPEND -> remoteKeyDao.loadPageInfo(subredditName)
    } 
    // query using page info
    db.withTransaction {
         db.withTransaction {
                if (loadType == REFRESH) {
                    postDao.deleteBySubreddit(subredditName)
                    remoteKeyDao.deleteBySubreddit(subredditName)
                }
                remoteKeyDao.insertOrRelace(SubredditPageInfo(name = subredditName, nextKey = response.nextKey)
                
         }
    }
}
```

Reply:
We only page forward so we only need append key right?

Reply:
Done